File: warehouse.py
"""


"""
import networkx as nx
import networkx.algorithms.shortest_paths.dense as nxalg
#import matplotlib.pyplot as plt

# Warehouse characteristics
LOCATION_X = 1
LOCATION_Y = 1
AISLES = 20
CROSS_AISLES = 1
AISLE_SIZE = 3
CROSS_AISLE_SIZE = 6
LOCATIONS = 10

# The graph instance
G = nx.Graph()

# Calculate cross points
cross_points = [0]
old = 0
for _ in range(CROSS_AISLES + 1):
    cross_points.append(old + LOCATIONS + 1)
    old += LOCATIONS + 1

# Calculate total nodes per side Y
locations_per_longaisle = LOCATIONS * (CROSS_AISLES + 1) + CROSS_AISLES + 2

# Populate the graph
node_id = 0
for x in range(AISLES):
    for y in range(locations_per_longaisle):
        G.add_node(node_id)
        if y > 0:
            weight = LOCATION_Y/2 + CROSS_AISLE_SIZE/2 if y in cross_points[1:-1] or (y - 1) in cross_points[1:-1] else LOCATION_Y
            G.add_edge(node_id, node_id - 1, weight=weight)
        if x > 0 and y in cross_points:
            G.add_edge(node_id, node_id - locations_per_longaisle, weight=LOCATION_X*2)
        node_id += 1

# Plot the graph
#nx.draw(G, with_labels=True, font_weight='bold')
#plt.show()

# Set the distance matrix
distance_matrix = nxalg.floyd_warshall_numpy(G)

File: src/warehouse/warehouse.go
package warehouse

// Define the warehouse characteristics
const LOCATION_X int = 2
const LOCATION_Y int = 2
const AISLES int = 20
const CROSS_AISLES int = 1
const AISLE_SIZE int = 3
const CROSS_AISLE_SIZE int = 6
const LOCATIONS int = 10

type GraphType map[int]map[int]int


func Graph() GraphType {
	// Initialize the graph
	var G GraphType = make(GraphType)
	// Calculate cross points
	cross_points := make(map[int]int)
	old := 0
	cross_points[0] = 1
	for i := 0; i < CROSS_AISLES + 1; i++{
		cross_points[old + LOCATIONS + 1] = 1
		old += LOCATIONS + 1
	}

	// Calculate total nodes per side Y
	locations_per_longaisle := LOCATIONS * (CROSS_AISLES + 1) + CROSS_AISLES + 2

	// Populate the graph
	nodeId := 0
	for x := 0; x < AISLES; x++ {
		for y := 0; y < locations_per_longaisle; y++ {
			// Calculate the connections
			var weight int
			if y > 0 {
				// Init the hashmap
				if _, ok := G[nodeId]; !ok { G[nodeId] = make(map[int]int) }
				if _, ok := G[nodeId - 1]; !ok { G[nodeId - 1] = make(map[int]int) }

				// Calculate distances
				weight = LOCATION_Y
				_, ok1 := cross_points[y]
				_, ok2 := cross_points[y - 1]
				if (ok1 && y != 0 && y != locations_per_longaisle) || (ok2 && y - 1 != 0 && y - 1 != locations_per_longaisle){
					weight = LOCATION_Y / 2 + CROSS_AISLE_SIZE / 2
				}
				G[nodeId][nodeId - 1] = weight
				G[nodeId - 1][nodeId] = weight
			}

			if _, ok := cross_points[y]; x > 0 && ok {
				// Init the hashmap
				if _, ok := G[nodeId]; !ok { G[nodeId] = make(map[int]int) }
				if _, ok := G[nodeId - locations_per_longaisle]; !ok { G[nodeId - locations_per_longaisle] = make(map[int]int) }
				// Calculate distances
				weight = LOCATION_X * 2
				G[nodeId][nodeId - locations_per_longaisle] = weight
				G[nodeId - locations_per_longaisle][nodeId] = weight
			}
			nodeId++
		}
	}
	return G
}



